### PR TITLE
Example Rule Template

### PR DESCRIPTION
**Pull request by horakj**
### Background

Customer requested a rule template to refer to like the policy template already in PA.

### Changes

* example_rule.py
* example_rule.yml

### Testing

* Nothing to test, just a generic template

### CODE REVIEW

File: templates/example_rule.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+AnalysisType: rule
+Filename: file_name.py 
+PolicyID: File.Name
```
Comment (lindsey-w) [inline templates/example_rule.yml:3]:
```suggestion
RuleID: File.Name
```

File: templates/example_rule.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+DisplayName: A More Friendly Name
```
Comment (lindsey-w) [inline templates/example_rule.yml:4]: I think we could also add:

```
OutputIds:
  - outputId
  - anotherOutputId
Suppressions:
  - suppression1
  - suppression2
```

File: templates/example_rule.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        return True
+    return False
+
+## Optional Functions
+def title(event): # User can set custom alert titles in string format
+    return 'This is my title'
+
+def dedup(event): # by default if not defined the deduplication period string is based off the alert title 
+    return 'DedupString'
+def alert_context(event): # additional information that the user may find useful that is passed in with the alert as a dictionary
```
Comment (lindsey-w) [inline templates/example_rule.py:13]:
```suggestion
    return 'DedupString'
    
def alert_context(event): # additional information that the user may find useful that is passed in with the alert as a dictionary
```

File: templates/example_rule.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    return dict(event)
+
+## Override Functions
+def severity(event): # change the severity of the events based which can be used if a special condition is met  (str = "INFO", "LOW", "MEDIUM', "HIGH", "CRITICAL")
+    return 'INFO'
+def description(event): # customize the description to the specific event (str)
```
Comment (lindsey-w) [inline templates/example_rule.py:19]:
```suggestion
    return 'INFO'
   
def description(event): # customize the description to the specific event (str)
```

File: templates/example_rule.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+AnalysisType: rule
```
Comment (nhakmiller) [inline templates/example_rule.yml:1]: Too lazy to convert it to YAML, but here are the list of fields you can use:

```
        "AnalysisType": Or("rule", "scheduled_rule"),
        "Enabled": bool,
        "Filename": str,
        "RuleID": And(str, NAME_ID_VALIDATION_REGEX),
        Or("LogTypes", "ScheduledQueries"): [str],
        "Severity": Or("Info", "Low", "Medium", "High", "Critical"),
        Optional("Description"): str,
        Optional("DedupPeriodMinutes"): int,
        Optional("DisplayName"): And(str, NAME_ID_VALIDATION_REGEX),
        Optional("OutputIds"): [str],
        Optional("Reference"): str,
        Optional("Runbook"): str,
        Optional("SummaryAttributes"): [str],
        Optional("Suppressions"): [str],
        Optional("Threshold"): int,
        Optional("Tags"): [str],
        Optional("Reports"): {str: list},
        Optional("Tests"): [
            {
                "Name": str,
                Optional(
                    "LogType"
                ): str,  # Not needed anymore, optional for backwards compatibility
                "ExpectedResult": bool,
                "Log": object,
            }
        ],
```

I would update this to indicate which ones are optional, and to include all the fields (e.g. I see `DedupPeriodMinutes` and `Threshold` are missing, important ones).

Or perhaps two template files, a minimal one with only required fields and a maximal one with all possible fields. They can both refer to the same python file.

**Review by lindsey-w — COMMENTED** [review]
Just had a couple of nits.  Thank you for creating this!

File: templates/example_rule.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+## Optional Functions
+def title(event): # User can set custom alert titles in string format
+    return 'This is my title'
+
+def dedup(event): # by default if not defined the deduplication period string is based off the alert title 
+    return 'DedupString'
+def alert_context(event): # additional information that the user may find useful that is passed in with the alert as a dictionary
+    return dict(event)
+
+## Override Functions
```
Comment (nhakmiller) [inline templates/example_rule.py:16]: There is no real distinction to Panther on what you label `Optional Functions` and `Override Functions`. In both cases they are using python to override YAML settings in the rule for a particular alert. We can distinguish here if you think that's useful, but functionally they're doing the same thing.

**Review by nhakmiller — APPROVED** [review]
Good to have. cc @glerb , interested to see what you think would be useful in a template coming from the other side.

File: templates/example_rule.yml (hunk 1 above)
Comment (horakj) [reply templates/example_rule.yml]: No problem, thanks for the catch

File: templates/example_rule.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+Filename: file_name.py 
+RuleID: File.Name
+DisplayName: A More Friendly Name
+Enabled: true
+ResourceTypes:
+  - Log.Type
```
Comment (nhakmiller) [inline templates/example_rule.yml:7]:
```suggestion
LogTypes:
  - Log.Type 
```

File: templates/example_rule.yml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  An optional Runbook 
+Reference: An optional reference.link
+SummaryAttributes:
+  - name
+  - action 
+Tests:
+  -
+    Name: Name 
+    ExpectedResult: false
+    Resource:
```
Comment (nhakmiller) [inline templates/example_rule.yml:26]:
```suggestion
    Log:
```

File: templates/example_rule.py (hunk 6 above)
Comment (horakj) [reply templates/example_rule.py]: Adding a Note at the the top to clarify this

File: templates/example_rule.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
+AnalysisType: rule # "rule" or "scheduled rule": str
+Filename: file_name.py # str
+RuleID: File.Name # str, NAME_ID_VALIDATION_REGEX),Or("LogTypes", "ScheduledQueries"): [str]
```
Comment (nhakmiller) [inline templates/example_rule.yml:3]:
```suggestion
RuleID: File.Name # str
```

File: templates/example_rule.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+DisplayName: A More Friendly Name # str, NAME_ID_VALIDATION_REGEX
```
Comment (nhakmiller) [inline templates/example_rule.yml:4]:
```suggestion
DisplayName: A More Friendly Name # str
```

File: templates/example_rule.yml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+Reports: # {str: list} (Optional)
+  Report Type:
+    - Report Name
+Severity: Medium # "Info", "Low", "Medium", "High", "Critical": str
+Description: > # str (Optional)
+  An optional Description
+Runbook: > # str (Optional)
+  An optional Runbook 
+Reference: An optional reference.link # str (Optional)
+DedupPeriodMinutes: 15 # int (Optional)
```
Comment (nhakmiller) [inline templates/example_rule.yml:22]:
```suggestion
DedupPeriodMinutes: 15 # int (Optional, defaults to 60)
```

File: templates/example_rule.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+Threshold:  5 # int (Optional)
```
Comment (nhakmiller) [inline templates/example_rule.yml:23]:
```suggestion
Threshold:  5 # int (Optional, defaults to 1)
```

**Review by nhakmiller — APPROVED** [review]
A few nits

**Pull request merged**